graphql_compiler/schema_generation/schema_graph.py

```diff
@@ -460,187 +430,231 @@ def _split_classes_by_kind(self, class_name_to_definition):
+            elem, links = _get_schema_element_and_links(class_name, class_name_to_definition,
+                                                        kind, inheritance_set,
+                                                        self._non_graph_class_names)
 
-        default_value = None
-        default_value_string = property_definition.get('defaultValue', None)
-        if default_value_string is not None:
-            default_value = parse_default_property_value(name, type_id, default_value_string)
+            # Freeze connections.
+            elem.freeze()
 
-        descriptor = PropertyDescriptor(type_id=type_id, qualifier=qualifier, default=default_value)
-        # Sanity-check the descriptor before returning it.
-        _validate_collections_have_default_values(class_name, name, descriptor)
-        return descriptor
+            # Perform validation.
+            _validate_links_not_defined_on_non_edge_class(class_name, kind, links)
 
-    def _link_vertex_and_edge_types(self):
-        """For each edge, link it to the vertex types it connects to each other."""
-        for edge_class_name in self._edge_class_names:
-            edge_element = self._elements[edge_class_name]
+            # Store.
+            self._elements[class_name] = elem
 
-            if (EDGE_SOURCE_PROPERTY_NAME not in edge_element.properties or
-                    EDGE_DESTINATION_PROPERTY_NAME not in edge_element.properties):
-                if edge_element.abstract:
-                    continue
-                else:
-                    raise AssertionError(u'Found a non-abstract edge class with undefined '
-                                         u'endpoint types: {}'.format(edge_element))
+    def _set_up_edge_elements(self, class_name_to_definition):
+        """Set up edge elements."""
+        kind = SchemaElement.ELEMENT_KIND_EDGE
+        for class_name in self.edge_class_names:
+            # Create edge SchemaElement.
+            inheritance_set = self._inheritance_sets[class_name]
+            elem, links = _get_schema_element_and_links(class_name, class_name_to_definition,
+                                                        kind, inheritance_set,
+                                                        self._non_graph_class_names)
+            abstract = _is_abstract(class_name_to_definition[class_name])
+
+            # Set edge connections. (Abstract classes may have missing leaf links).
+            in_leaf_link = _try_get_leaf_link(
+                class_name, EDGE_SOURCE_PROPERTY_NAME, links, abstract, self._subclass_sets)
+            out_leaf_link = _try_get_leaf_link(
+                class_name, EDGE_DESTINATION_PROPERTY_NAME, links, abstract, self._subclass_sets)
+            if in_leaf_link is not None:
+                elem.in_connections.add(in_leaf_link)
+            if out_leaf_link is not None:
+                elem.out_connections.add(out_leaf_link)
```

Comment:
Do we only need these leaf links when the edge is not abstract? Can we not call `_try_get_leaf_link` on abstract edges at all? That will make the function spec and the usage much simpler.

Reply:
Nope we need for abstract classes too.

Reply:
Actually we need it almost solely for abstract edges. The leaf links of a non-abstract edge are just the links that were defined in it.